lib/oli/activities/transformers.ex

```diff
@@ -1,46 +1,161 @@
+    model_repository = Enum.reduce(revisions, %{}, fn r, m -> Map.put(m, r.id, r.content) end)
+
+    # Parse the models, and create a list of tuples of the revision and the parse result
+    parsed_models = Enum.map(revisions, fn r -> {r, Model.parse(r.content)} end)
+
+    parsing_errors_by_id = index_parsing_errors(parsed_models)
+
+    # Create a MapSet of all revision ids that parsed successfully and that have at least one transformation
+    has_transformation =
+      Enum.filter(parsed_models, fn {_, parsed_model_result} ->
+        case parsed_model_result do
+          {:ok, parsed_model} -> Enum.count(parsed_model.transformations) > 0
+          _ -> false
         end
+      end)
+      |> Enum.map(fn {rev, _} -> rev.id end)
+      |> MapSet.new()
+
+    # Group the transformers by type and batch execute their context creation and individually
+    # apply their transforms
+    {model_repository, transform_errors} =
+      group_by_operation(parsed_models)
+      |> batch_execute(model_repository)
 
-      {:error, error} ->
-        {:error, error}
+    # Now map our original list of revisions to the result of the transformation.
+    Enum.map(revisions, fn r ->
+      if MapSet.member?(has_transformation, r.id) do
+        if Map.has_key?(transform_errors, r.id) do
+          {:error, Map.get(transform_errors, r.id)}
+        else
+          {:ok, Map.get(model_repository, r.id)}
+        end
+      else
+        if Map.has_key?(parsing_errors_by_id, r.id) do
+          {:error, Map.get(parsing_errors_by_id, r.id)}
+        else
+          {:ok, nil}
+        end
+      end
+    end)
+  end
+
+  # Executes all transformations, grouped by operation (aka, their "type"), in a manner
+  # that gathers any required context in a batch manner.
+  #
+  # Takes a map of operations as keys, with values being a list of tuples of revision id
+  # and transformation.  Second argument is a map of revision models, keyed by their revision
+  # id.
+  #
+  # Returns a tuple with the first element being the updated map of revision models and the
+  # second element being a map of revision ids to errors encountered.
+  defp batch_execute(by_operation, model_repository) do
+    @supported_transformations_by_operation
+    |> Enum.reduce({model_repository, %{}}, fn {operation, module},
+                                               {model_repository, errors_by_id} ->
+      case Map.get(by_operation, operation) do
+        nil ->
+          {model_repository, errors_by_id}
+
+        revision_id_transform_pairs ->
+          run_one_operation(model_repository, module, revision_id_transform_pairs, errors_by_id)
+      end
+    end)
+  end
+
+  # Run one batch of transformers, for a specific operation.
+  #
+  # Takes the model repository, the module of the transformer impl, and a list of
+  # tuples of revision ids and transformer.  The larevision_id_transform_pairsst argument is really our
```

Comment:
Should this say, "The revision_id_transform_pairs argument ..."?

Reply:
It indeed should.  Changes pushed